day-one-challenge/tyranny-of-the-rocket/getFuelRequirements.js

```diff
@@ -0,0 +1,20 @@
+const { sum, map } = require('ramda');
+const moduleMassArray = require('./fuelData');
+
+const calculateFuel = mass => {
+  return Math.floor(mass / 3) - 2;
+};
+
+const fuelOfCurrentModuleOfMass = (currentMass) => {
+  if (typeof (currentMass) !== 'number')
+    throw new TypeError("mass entered is not a number!");
+  return calculateFuel(currentMass) <= 0 ? 0 : calculateFuel(currentMass) + fuelOfCurrentModuleOfMass(calculateFuel(currentMass));
```

Comment:
So, this function is not tail-call optimized yet, do you remember how to tell?

Also, you and I found out something interesting about node(V8) and its ability to take advantage of TCO code. Do you mind elaborating?

Reply:
A function is tail-call optimized if it does not need to hold onto information from previous function calls. A tail-call optimized function passes along, via parameters, all the information needed for calling the function so the call stack should not grow each time the function is called. Here is the implementation of TCO we worked on when we paired: 99480b585e4036f563197c29bb96af1cadeebb1a

We found out that node stopped supporting TCO as of version 8.x, and even though a function may be tail call optimized, node does not recognize this and the call stack will continue to grow.